File: Recursion/Leetcode/Easy/Fibonacci/Fibonacci.cpp
#include<bits/stdc++.h>
using namespace std;
long nthFibo(long n, map<long, long> &hash) {
    if(n == 0) {
        return 0;
    } else if(hash[n]) {
        return hash[n];
    }
    hash[n] = nthFibo(n - 1, hash) + nthFibo(n - 2, hash);
    return hash[n];
}
long fibonacci(long n) {
    map<long, long> hash;
    hash[0] = 0;
    hash[1] = 1;
    return nthFibo(n, hash);
}
int main(void) {
    cout << fibonacci(30);
    return 0;
}

File: Recursion/Leetcode/Easy/Fibonacci/Fibonacci.java
import java.util.HashMap;

public class Fibonacci {

    static int nthFibo(int n, HashMap<Integer, Integer> hash) {
        if (n == 0) {
            return 0;
        }
        int x = hash.getOrDefault(n, -1);
        if(x != -1) {
            return x;
        }
        x = nthFibo(n - 1, hash) + nthFibo(n - 2, hash);
        hash.put(n, x);
        return x;
    }

    static int fibonacci(int n) {
        HashMap<Integer, Integer> hash = new HashMap<>();
        hash.put(0, 0);
        hash.put(1, 1);
        return nthFibo(n, hash);
    }

    public static void main(String[] args) {
        System.out.println(fibonacci(15));
    }
}
